pkg/resourcemanager/mysql/mysqlhelper.go

```diff
@@ -88,30 +93,112 @@ func ConnectToSQLDBAsCurrentUser(
+	if err := helpers.FindBadChars(user); err != nil {
+		return nil, errors.Wrapf(err, "problem found with username")
+	}
+
+	// TODO: Possibly need to fully qualify this with database name?
+	rows, err := db.QueryContext(ctx, "SHOW GRANTS FOR ?", user)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to list grants for user %s", user)
+	}
+	defer rows.Close()
+
+	result := make(map[string]map[string]struct{})
+	for rows.Next() {
+		var row string
+		err := rows.Scan(&row)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to iterate returned rows")
+		}
+
+		grants, db, err := parseGrantsRow(row)
+		if err != nil {
+			return nil, err
+		}
+
+		result[db] = grants
+	}
+
+	return result, nil
+}
+
 func GrantUserRoles(ctx context.Context, user string, database string, roles []string, db *sql.DB) error {
 	var errorStrings []string
 	if err := helpers.FindBadChars(user); err != nil {
 		return fmt.Errorf("problem found with username: %v", err)
 	}
 
+	rolesMap := make(map[string]struct{})
 	for _, role := range roles {
+		rolesMap[role] = struct{}{}
+	}
+
+	// Get the current roles
+	currentRoles, err := ExtractUserRoles(ctx, user, db)
+	if err != nil {
+		return errors.Wrapf(err, "couldn't get existing roles for user %s", user)
+	}
+	currentRolesForDB := currentRoles[strings.ToLower(database)]
 
-		if err := helpers.FindBadChars(role); err != nil {
+	// Remove "USAGE" as it's special and we never grant or remove it
+	delete(currentRolesForDB, "USAGE")
+
+	rolesAndActions := helpers.DiffCurrentAndExpectedSQLRoles(currentRolesForDB, rolesMap)
+
+	for _, roleAndAction := range rolesAndActions {
+
+		if err := helpers.FindBadChars(roleAndAction.Role); err != nil {
 			return fmt.Errorf("problem found with role: %v", err)
 		}
 
 		// Due to how go-mysql-driver performs parameter replacement, it always wraps
-		// string parameters in ''. That doesn't work for this query because some of
-		// our parameters are actually SQL keywords or identifiers (backticks). Admittedly
+		// string parameters in ''. That doesn't work for these queries because some of
+		// our parameters are actually SQL keywords or identifiers (requiring backticks). Admittedly
 		// protecting against SQL injection here is probably pointless as we're giving the caller
 		// permission to create users, which means there's nothing stopping them from creating
 		// an administrator user and then doing whatever they want without SQL injection.
 		// See https://github.com/go-sql-driver/mysql/blob/3b935426341bc5d229eafd936e4f4240da027ccd/connection.go#L198
 		// for specifics of what go-mysql-driver supports.
-		tsql := fmt.Sprintf("GRANT %s ON `%s`.* TO ?", role, database)
-		_, err := db.ExecContext(ctx, tsql, user)
-		if err != nil {
-			errorStrings = append(errorStrings, err.Error())
+		switch roleAndAction.Action {
+		case helpers.RoleActionCreate:
+			tsql := fmt.Sprintf("GRANT %s ON `%s`.* TO ?", roleAndAction.Role, database)
```

Comment:
The GRANT and REVOKE commands can accept multiple roles, so you could just issue two statements (or one if one of the slices was empty) rather than one per role changed.

Reply:
Nice